indico/client/request.py

```diff
@@ -13,41 +21,55 @@ class HTTPMethod(Enum):
     OPTIONS = "OPTIONS"
 
 
-class HTTPRequest:
-    def __init__(self, method: HTTPMethod, path: str, **kwargs):
-        self.method = method
-        self.path = path
-        self.kwargs = kwargs
+class HTTPRequest(Generic[ResponseType]):
+    def __init__(self, method: HTTPMethod, path: str, **kwargs: "Any"):
+        self.method: HTTPMethod = method
+        self.path: str = path
+        self._kwargs: "AnyDict" = kwargs
+
+    @property
+    def kwargs(self) -> "AnyDict":
+        return self._kwargs
+
+    def process_response(self, response: "Any") -> "ResponseType":
+        return cast("ResponseType", response)
 
-    def process_response(self, response):
-        return response
 
+class GraphQLRequest(Generic[ResponseType], HTTPRequest[ResponseType]):
+    def __init__(self, query: str, variables: "Optional[AnyDict]" = None):
+        self.query: str = query
+        self.variables: "Optional[AnyDict]" = variables
 
-class GraphQLRequest(HTTPRequest):
-    def __init__(self, query: str, variables: Dict[str, Any] = None):
-        self.query = query
-        self.variables = variables
-        self.method = HTTPMethod.POST
-        self.path = "/graph/api/graphql"
+        super().__init__(HTTPMethod.POST, "/graph/api/graphql")
 
     @property
-    def kwargs(self):
+    def kwargs(self) -> "AnyDict":
         return {"json": {"query": self.query, "variables": self.variables}}
 
-    def process_response(self, response):
-        response = super().process_response(response)
-        errors = response.pop("errors", [])
+    def parse_payload(self, response: "AnyDict") -> "Any":
+        raw_response: "AnyDict" = cast("AnyDict", super().process_response(response))
+        errors: "List[AnyDict]" = raw_response.pop("errors", [])
+
         if errors:
-            extras = {"locations": [error.pop("locations", None) for error in errors]}
+            extras: "Dict[str, List[Any]]" = {
+                "locations": [error.pop("locations", None) for error in errors]
+            }
+
             raise IndicoRequestError(
                 error="\n".join(error["message"] for error in errors),
                 code=400,
                 extras=extras,
             )
-        return response["data"]
 
+        return raw_response["data"]
+
+    def process_response(self, response: "AnyDict") -> "ResponseType":
```

Comment:
typed python does not permit overriding the typing of the signature in subclasses. we were doing that implicitly before by having this class return `Dict` and subclasses return their `BaseType`, etc. 

to avoid having to type every single use of `super().process_response` with a `cast()` or `type: ignore`, i split the function into a `parse_payload` typed to return `Any`.

practically, that means we're deprecating the use of `super().process_response` in any custom `GraphQLRequest` subclasses; i tried, but could not figure out a nice way to raise a deprecation error so folks know to switch over

Reply:
We've used https://deprecation.readthedocs.io/en/latest/  in the past for raising deprecation errors

Reply:
yea, i just couldn't figure out a way to _only_ raise the error when invoking it via `super().process_response`

Reply:
my kingdom for working deprecation in python